src/mlpack/base.hpp

```diff
@@ -0,0 +1,124 @@
+/**
+ * @file base.hpp
+ *
+ * The most basic core includes that mlpack expects; standard C++ includes and
+ * Armadillo only
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_BASE_HPP
+#define MLPACK_BASE_HPP
+
+// First, check if Armadillo was included before, warning if so.
+#ifdef ARMA_INCLUDES
+#pragma message "Armadillo was included before mlpack; this can sometimes cause\
+ problems.  It should only be necessary to include <mlpack/core.hpp> and not \
+<armadillo>."
+#endif
+
+// Defining _USE_MATH_DEFINES should set M_PI.
+#define _USE_MATH_DEFINES
+#include <cmath>
+
+// Next, standard includes.
+#include <cctype>
+#include <cfloat>
+#include <climits>
+#include <cstdint>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <stdexcept>
+#include <tuple>
+#include <utility>
+#include <numeric>
+#include <vector>
+#include <queue>
+
+// But if it's not defined, we'll do it.
+#ifndef M_PI
+  #define M_PI 3.141592653589793238462643383279
+#endif
+
+// MLPACK_COUT_STREAM is used to change the default stream for printing
+// purpose.
+#if !defined(MLPACK_COUT_STREAM)
+ #define MLPACK_COUT_STREAM std::cout
+#endif
+
+// MLPACK_CERR_STREAM is used to change the stream for printing warnings
+// and errors.
+#if !defined(MLPACK_CERR_STREAM)
+ #define MLPACK_CERR_STREAM std::cerr
+#endif
+
+// Give ourselves a nice way to force functions to be inline if we need.
+#define force_inline
+#if defined(__GNUG__) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __attribute__((always_inline))
+#elif defined(_MSC_VER) && !defined(DEBUG)
+  #undef force_inline
+  #define force_inline __forceinline
+#endif
+
+// Backport this functionality from C++14, if it doesn't exist.
+#if __cplusplus <= 201103L
+#if !defined(_MSC_VER) || _MSC_VER <= 1800
+namespace std {
```

Comment:
Adding stuff to the `std` namespace is risky

Reply:
Thanks for pointing this out---this little bit got missed in the discussion in #2939.  I also noticed that the `FORCE_CXX11` CMake option can be removed too using this reasoning (66923e8782), and then I removed this check too in 2fcec24991.